### PR TITLE
Hotfix/dynamic validators2

### PR DESCRIPTION
**Pull request by MRamanenkau**
### Types of Changes

- [ ] Tech Debt (Code improvements)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Dependency upgrade (A change in substrate or any 3rd party crate version)

### Migrations and Hooks

- [ ] This change requires a runtime migration.
- [ ] Modifies `on_initialize`
- [ ] Modifies `on_finalize`

### Checklist

- [ ] Change has been tested locally.
- [ ] Change adds / updates tests.
- [ ] Changelog doc updated.

### CODE REVIEW

File: pallets/validators/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::add_one())]
+		pub fn add_one(origin: OriginFor<T>, validator_to_add: T::AccountId) -> DispatchResult {
+			ensure_root(origin)?;
+
+			let mut validators = Validators::<T>::get();
+			ensure!(!validators.contains(&validator_to_add), Error::<T>::AlreadyExists);
+			validators.push(validator_to_add.clone());
+			Validators::<T>::put(validators);
```
Comment (rakanalh) [inline pallets/validators/src/lib.rs:70]: I think we can use `try_mutate` here for the two steps

File: pallets/validators/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+		pub fn remove_one(
+			origin: OriginFor<T>,
+			validator_to_remove: T::AccountId,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+			let mut validators = Validators::<T>::get();
+			ensure!(validators.contains(&validator_to_remove), Error::<T>::DoNotExists);
+
+			let index = validators.iter().position(|v| *v == validator_to_remove).unwrap();
+			validators.remove(index);
```
Comment (rakanalh) [inline pallets/validators/src/lib.rs:88]: Same here, we can do it in one step i think.

File: pallets/validators/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+	pub enum Error<T> {
+		AlreadyExists,
+		DoNotExists,
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn validators)]
+	pub(crate) type Validators<T: Config> = StorageValue<_, Vec<T::AccountId>, ValueQuery>;
+
+	#[pallet::call]
```
Comment (rakanalh) [inline pallets/validators/src/lib.rs:60]: Please restore the batch `set_validators` to override all.

**Review by rakanalh — CHANGES_REQUESTED** [review]

File: pallets/validators/src/lib.rs (hunk 3 above)
Comment (MRamanenkau) [reply pallets/validators/src/lib.rs:62]: Restored

File: pallets/validators/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+			origin: OriginFor<T>,
+			validator_to_remove: T::AccountId,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+			let mut validators = Validators::<T>::get();
+			ensure!(validators.contains(&validator_to_remove), Error::<T>::DoNotExists);
+
+			Validators::<T>::try_mutate(|validators| -> DispatchResult {
+				let index = validators.iter().position(|v| *v == validator_to_remove).ok_or(Error::<T>::DoNotExists)?;
+				validators.remove(index);
```
Comment (rakanalh) [inline pallets/validators/src/lib.rs:92]: No need to `let index` line.

```
validators.retain(|&account| account != validator_to_remove);
```

File: pallets/validators/src/lib.rs (hunk 1 above)
Comment (MRamanenkau) [reply pallets/validators/src/lib.rs]: Updated

File: pallets/validators/src/lib.rs (hunk 2 above)
Comment (MRamanenkau) [reply pallets/validators/src/lib.rs]: Updated

File: pallets/validators/src/lib.rs (hunk 3 above)
Comment (MRamanenkau) [reply pallets/validators/src/lib.rs:62]: Updated

File: pallets/validators/src/lib.rs (hunk 4 above)
Comment (MRamanenkau) [reply pallets/validators/src/lib.rs]: Updated

**Pull request closed without merging**